bytes_diff_visualizer/bytes_diff_visualizer.go

```diff
@@ -82,6 +91,30 @@ func formatBytesAsAscii(value []byte) string {
 	return strings.Join(asciiRepresentations, "")
 }
 
+func formatHexWithColorizedByte(value []byte, i int, firstDiffIndex int, end int, chosenColor color.Attribute) string {
+	if firstDiffIndex >= i && firstDiffIndex < end {
+		return PadRight(formatHexWithColorizedByteHelper(value, i, firstDiffIndex, end, chosenColor), " ", 69)
+	} else {
+		return PadRight(formatBytesAsHex(value[i:end]), " ", 60)
+	}
+}
+
+func formatHexWithColorizedByteHelper(value []byte, i int, firstDiffIndex int, end int, chosenColor color.Attribute) string {
+	return formatBytesAsHex(value[i:firstDiffIndex]) + " " + colorizeString(chosenColor, formatBytesAsHex(value[firstDiffIndex:firstDiffIndex+1])) + " " + formatBytesAsHex(value[firstDiffIndex+1:end])
+}
+
+func formatAsciiWithColorizedByte(value []byte, i int, firstDiffIndex int, end int, chosenColor color.Attribute) string {
+	if firstDiffIndex >= i && firstDiffIndex < end {
+		return formatAsciiWithColorizedByteHelper(value, i, firstDiffIndex, end, chosenColor)
+	} else {
+		return formatBytesAsAscii(value[i:end])
+	}
+}
+
+func formatAsciiWithColorizedByteHelper(value []byte, i int, firstDiffIndex int, end int, chosenColor color.Attribute) string {
```

Comment:
@ryan-gang there's still a bunch of complexity here around byte indexes - the main reason for all of this seems to be that we're using `formatBytesAsHex` instead of `formatByteAsHex`. Would things be simpler if we were printing out byte-by-byte instead of constructing a whole line at once and then having to mutate it?

This is certainly cleaner than what we were doing to the TabWriter output, but I think it can be far simpler. Good enough to merge, but I'd suggest giving this another go if you're looking to practice code design skills - good opportunity!

Reply:
Merging this for now, will work on the other implementation separately. 
We can actually inline all the hex and ASCII conversion, to make this simpler, that way we can go byte-by-byte as you mentioned.